### ISSUE TITLE
A lightweight implementation that dramatically improves runtime of xicor

### ISSUE THREAD

**Issue by atarashansky** [czbiohub/pyxi#17]
Hello, I wrote a lightweight (and much faster) implementation of the xi correlation metric.

In my tests, this implementation is > 300x faster than the current implementation.

```python
import numpy as np

def _xicorr(X,Y):
    n = X.size
    xi = np.argsort(X,kind='quicksort')
    Y = Y[xi]
    _,b,c = np.unique(Y,return_counts=True,return_inverse=True)
    r = np.cumsum(c)[b]
    _,b,c = np.unique(-Y,return_counts=True,return_inverse=True)
    l = np.cumsum(c)[b]
    return 1 - n*np.abs(np.diff(r)).sum() / (2*(l*(n-l)).sum())
```

Hopefully this is helpful!

**Comment by j-bac** [comment]
@atarashansky 

There's some issues with your proposed function

```python
np.random.seed(0)
x=np.random.random((100,1))
y=np.sin(x)
_xicorr(y,x)
```
Returns nan, while Xi.corr(x,y) return 0.9997

**Comment by j-bac** [comment]
@atarashansky 
This is very helpful ! Here's a further 5x speedup using numba:

```python
import numba as nb
@nb.njit
def nb_unique1d(ar):
    """
    Find the unique elements of an array, ignoring shape.
    """
    ar = ar.flatten()

    optional_indices = return_index or return_inverse

    if optional_indices:
        perm = ar.argsort(kind='mergesort')
        aux = ar[perm]
    else:
        ar.sort()
        aux = ar
    mask = np.empty(aux.shape, dtype=np.bool_)
    mask[:1] = True
    if aux.shape[0] > 0 and aux.dtype.kind in "cfmM" and np.isnan(aux[-1]):
        if aux.dtype.kind == "c":  # for complex all NaNs are considered equivalent
            aux_firstnan = np.searchsorted(np.isnan(aux), True, side='left')
        else:
            aux_firstnan = np.searchsorted(aux, aux[-1], side='left')
        mask[1:aux_firstnan] = (aux[1:aux_firstnan] != aux[:aux_firstnan - 1])
        mask[aux_firstnan] = True
        mask[aux_firstnan + 1:] = False
    else:
        mask[1:] = aux[1:] != aux[:-1]


    imask = np.cumsum(mask) - 1
    inv_idx = np.empty(mask.shape, dtype=np.intp)
    inv_idx[perm] = imask
    idx = np.append(np.nonzero(mask)[0],mask.size)

                     #idx      #inverse   #counts
    return aux[mask],perm[mask],inv_idx,np.diff(idx)

@nb.njit
def _xicorr_v2(X,Y):
    '''xi correlation coefficient'''
    n = X.size
    xi = np.argsort(X,kind='quicksort')
    Y = Y[xi]
    _,_,b,c = nb_unique1d(Y)
    r = np.cumsum(c)[b]
    _,_,b,c = nb_unique1d(-Y)
    l = np.cumsum(c)[b]
    return 1 - n*np.abs(np.diff(r)).sum() / (2*(l*(n-l)).sum())
```

**Comment by atarashansky** [comment]
Brilliant! I didn't think to reimplement the unique function in numba. Thanks so much for the improvement!

Is the `nan` issue still present?

**Comment by j-bac** [comment]
The nan issue was my bad, I was providing x and y with shape (n_obs,1) rather than (n_obs,) ! 
I work on genomics sequencing data with many variables so I also wrote this to loop over many variables efficiently. If you know numba it is straightforward with `prange` but happy to share if you want. 

I would be interested to know if you invented an efficient version for the other functionalities like p-values, otherwise I will take a look at it also.

**Comment by atarashansky** [comment]
Please, do share the `prange` implementation! I've been using multiprocessing pools to parallelize as there are certain elements of my code that cannot be written in no-python numba, but it would be nice to have the `prange` approach for other applications.

**Comment by escherba** [comment]
I didn't have time to dig into the math, but there is a discrepancy between @atarashansky implementation and original implementation in R published here https://github.com/cran/XICOR/blob/master/R/calculateXI.R. Example:
```python
def test_Xi_01():
    x = np.array([2.0, 5.0, 8.0, 1.0, 3.0, 1.0])
    y = np.array([7.0, 5.0, 3.0, 6.0, 7.0, 9.0])
    result = xicorr(x, y)
    expected = 0.3239437  # alternative implementation returns 0.3333333
    assert_almost_equal(result, expected)
```
Relatedly, I translated the R implementation into Python, and it also seems to be very fast. Here is the Python translation:
```python
def xicorr(x: 'ArrayLike', y: 'ArrayLike') -> float:
    """Original R implementation of Xi translated into Python

    R implementation:
    https://github.com/cran/XICOR/blob/master/R/calculateXI.R
    """
    x = np.asarray(x)
    y = np.asarray(y)
    n = x.size
    assert y.size == n, "arrays must be of the same size"
    PI = rankdata(x, method="average")
    fr = rankdata(y, method="average")
    CU = np.mean((float(n + 1) - fr) * (fr - 1.))
    A1 = np.abs(np.diff(fr[np.argsort(PI, kind="quicksort")])).sum()
    return 1. - 0.5 * float(A1 / CU)
```

**Comment by escherba** [comment]
There is additional speed improvement to be had if we call `np.unique` only once:

```python
import numpy as np

def xicorr(x: 'ArrayLike', y: 'ArrayLike') -> float:
    """xi correlation coefficient (alternative implementation)

    Written by github/atarashansky, modified by github/escherba
    https://github.com/czbiohub/xicor/issues/17#issue-965635013
    """
    x = np.asarray(x)
    y = np.asarray(y)
    n = x.size
    assert y.size == n, "arrays must be of the same size"
    y = y[np.argsort(x, kind='quicksort')]
    _, inverse, counts = np.unique(y, return_inverse=True, return_counts=True)
    right = np.cumsum(counts)[inverse]
    left = np.cumsum(np.flip(counts))[(counts.size - 1) - inverse]
    return 1. - 0.5 * float(np.abs(np.diff(right)).sum() / np.mean(left * (n - left)))
```

By the way, the original R implementation, translated into Python, is also quite fast:
```python
from scipy.stats import rankdata

def xicorr_orig(x: 'ArrayLike', y: 'ArrayLike') -> float:
    """Original R implementation of Xi translated into Python

    R implementation:
    https://github.com/cran/XICOR/blob/master/R/calculateXI.R
    """
    x = np.asarray(x)
    y = np.asarray(y)
    n = x.size
    assert y.size == n, "arrays must be of the same size"
    PI = rankdata(x, method="average")
    fr = rankdata(y, method="average")
    CU = np.mean((float(n + 1) - fr) * (fr - 1.))
    A1 = np.abs(np.diff(fr[np.argsort(PI, kind="quicksort")])).sum()
    return 1. - 0.5 * float(A1 / CU)
```

**Comment by rsdouglas** [comment]
@escherba yours is the only implementation that works on my data. With everything else I'm getting the following exception:
```TypeError: bad operand type for unary -: 'str'```
Which is a result of `-Y` here:
```    _,b,c = np.unique(-Y,return_counts=True,return_inverse=True)```

**Comment by AbhiN2001** [comment]
> Hello, I wrote a lightweight (and much faster) implementation of the xi correlation metric.
> 
> In my tests, this implementation is > 300x faster than the current implementation.
> 
> ```python
> import numpy as np
> 
> def _xicorr(X,Y):
>     n = X.size
>     xi = np.argsort(X,kind='quicksort')
>     Y = Y[xi]
>     _,b,c = np.unique(Y,return_counts=True,return_inverse=True)
>     r = np.cumsum(c)[b]
>     _,b,c = np.unique(-Y,return_counts=True,return_inverse=True)
>     l = np.cumsum(c)[b]
>     return 1 - n*np.abs(np.diff(r)).sum() / (2*(l*(n-l)).sum())
> ```
> 
> Hopefully this is helpful!

How does this implementation break ties at random?

**Comment by janarez** [comment]
> > ```python
> > import numpy as np
> > 
> > def _xicorr(X,Y):
> >     n = X.size
> >     xi = np.argsort(X,kind='quicksort')
> >     Y = Y[xi]
> >     _,b,c = np.unique(Y,return_counts=True,return_inverse=True)
> >     r = np.cumsum(c)[b]
> >     _,b,c = np.unique(-Y,return_counts=True,return_inverse=True)
> >     l = np.cumsum(c)[b]
> >     return 1 - n*np.abs(np.diff(r)).sum() / (2*(l*(n-l)).sum())
> > ```
>
> How does this implementation break ties at random?

@AbhiN2001 Hi, it doesn't break ties uniformly at random. If you have duplicates you need to add random noise for the `argsort`, e.g.: 
```python
(X + np.random.rand(len(X))*(10**(-6))).argsort(kind='quicksort')
```

**Comment by rpeck** [comment]
> > How does this implementation break ties at random?
> 
> @AbhiN2001 Hi, it doesn't break ties uniformly at random. If you have duplicates you need to add random noise for the `argsort`, e.g.:
> 
> ```python
> (X + np.random.rand(len(X))*(10**(-6))).argsort(kind='quicksort')
> ```

To state the obvious (one of my strengths), if the X values are very small or have a very limited range (if either are near 10^-6) this solution is not a good one…

**Comment by atarashansky** [comment]
> There is additional speed improvement to be had if we call `np.unique` only once:
> 
> ```python
> import numpy as np
> 
> def xicorr(x: 'ArrayLike', y: 'ArrayLike') -> float:
>     """xi correlation coefficient (alternative implementation)
> 
>     Written by github/atarashansky, modified by github/escherba
>     https://github.com/czbiohub/xicor/issues/17#issue-965635013
>     """
>     x = np.asarray(x)
>     y = np.asarray(y)
>     n = x.size
>     assert y.size == n, "arrays must be of the same size"
>     y = y[np.argsort(x, kind='quicksort')]
>     _, inverse, counts = np.unique(y, return_inverse=True, return_counts=True)
>     right = np.cumsum(counts)[inverse]
>     left = np.cumsum(np.flip(counts))[(counts.size - 1) - inverse]
>     return 1. - 0.5 * float(np.abs(np.diff(right)).sum() / np.mean(left * (n - left)))
> ```
> 
> By the way, the original R implementation, translated into Python, is also quite fast:
> 
> ```python
> from scipy.stats import rankdata
> 
> def xicorr_orig(x: 'ArrayLike', y: 'ArrayLike') -> float:
>     """Original R implementation of Xi translated into Python
> 
>     R implementation:
>     https://github.com/cran/XICOR/blob/master/R/calculateXI.R
>     """
>     x = np.asarray(x)
>     y = np.asarray(y)
>     n = x.size
>     assert y.size == n, "arrays must be of the same size"
>     PI = rankdata(x, method="average")
>     fr = rankdata(y, method="average")
>     CU = np.mean((float(n + 1) - fr) * (fr - 1.))
>     A1 = np.abs(np.diff(fr[np.argsort(PI, kind="quicksort")])).sum()
>     return 1. - 0.5 * float(A1 / CU)
> ```
> 
> I previously posted about a discrepancy I observed, but it seems to have been caused by a tie found in the Y array. Ignore that previous message.

1) Does the R implementation translated into python break ties at random? I'm not sure how `rankdata` handles it internally. 
2) What's the speed compared to the numba implementation?

**Comment by atarashansky** [comment]
> > > How does this implementation break ties at random?
> > 
> > 
> > @AbhiN2001 Hi, it doesn't break ties uniformly at random. If you have duplicates you need to add random noise for the `argsort`, e.g.:
> > ```python
> > (X + np.random.rand(len(X))*(10**(-6))).argsort(kind='quicksort')
> > ```
> 
> To state the obvious (one of my strengths), if the X values are very small or have a very limited range (if either are near 10^-6) this solution is not a good one…

The noise can be as small as we have precision for, so maybe the solution is to just do something like scale the noise level by the mean of values in `X`
```python
(X + np.random.rand(len(X))*X.mean()*(10**(-6))).argsort(kind='quicksort')
```
Or even simpler just make the noise so tiny that the failure mode is exceedingly unlikely
```python
(X + np.random.rand(len(X))*(10**(-20))).argsort(kind='quicksort') #assuming we have enough precision, probably need a dtype check or something.
```

**Comment by rpeck** [comment]
My overall point is that this sort of detail is what makes numerical methods so difficult to get right for all cases. E.g., rather than simply scale by the mean, one should look at the range to ensure that the random values we're adding are much smaller than the range. You might have a list of large numbers that have a small relative range. I've seen this in real datasets (IIRC, it was in a scientific dataset). I haven't thought through what the best solution is here, but these things can be subtle.

One of the interview questions I often give to DS people is, Given a long but arbitrary list of floating point numbers, how can we most accurately add them up and why can this be an issue? Most people haen't thought about the implications of floating point underflow on adding tiny numbers to large ones, or on subtracting numbers that are very close in their mantissa.

If we're adding rand()*10^-20 to large values the mantissa is likely to get shifted to zero in the addition.

**Comment by rpeck** [comment]
The most-correct solution, IMO, would be to take a single pass through X to identify spans of identical values and then shuffle the corresponding spans of Y.

**Comment by rpeck** [comment]
That solution also works for categorical X.

Now, being lazy, I'll wait for someone else to implement proper duplicate handling. :-)
